tests/test_api.py

```diff
@@ -0,0 +1,332 @@
+    order = await create_test_order(set_up)
+
+    # Cancel the created order and ensure it's cancelled
+    cancelled_order = await set_up["client"].cancel_gridpool_order(
+        GRIDPOOL_ID, order.order_id
+    )
+    assert cancelled_order.order_id == order.order_id, "Order cancellation failed"
+
+    fetched_order = await set_up["client"].get_gridpool_order(
+        GRIDPOOL_ID, order.order_id
+    )
+    assert (
+        fetched_order.state_detail.state == OrderState.CANCELED
+    ), "Order state should be CANCELED"
+
+
+async def test_update_cancelled_order_failure(set_up: dict[str, Any]) -> None:
+    """Test updating a cancelled order and ensure it fails."""
+    # Create an order first
+    order = await create_test_order(set_up)
+
+    # Cancel the created order
+    await set_up["client"].cancel_gridpool_order(GRIDPOOL_ID, order.order_id)
+
+    # Expected failure as cancelled order cannot be updated
+    with pytest.raises(grpc.aio.AioRpcError) as excinfo:
+        await set_up["client"].update_gridpool_order(
+            gridpool_id=GRIDPOOL_ID, order_id=order.order_id, price=set_up["price"]
+        )
+    assert (
+        excinfo.value.code() == grpc.StatusCode.INVALID_ARGUMENT
+    ), "Expected INVALID_ARGUMENT error"
+
+
+async def test_cancel_all_orders(set_up: dict[str, Any]) -> None:
+    """Test cancelling all orders."""
+    # Create multiple orders
+    for _ in range(10):
+        await create_test_order(set_up)
+
+    # Cancel all orders and check that did indeed get cancelled
+    await set_up["client"].cancel_all_gridpool_orders(GRIDPOOL_ID)
+
+    orders = await set_up["client"].list_gridpool_orders(gridpool_id=GRIDPOOL_ID)
+
+    for order in orders:
+        assert (
+            order.state_detail.state == OrderState.CANCELED
+        ), f"Order {order.order_id} not canceled"
+
+
+async def test_list_gridpool_trades(set_up: dict[str, Any]) -> None:
+    """Test listing gridpool trades."""
+    buy_order, sell_order = await create_test_trade(set_up)
+    trades = await set_up["client"].list_gridpool_trades(
+        GRIDPOOL_ID,
+        delivery_period=buy_order.order.delivery_period,
+    )
+    assert len(trades) >= 1
+
+
+async def test_list_public_trades(set_up: dict[str, Any]) -> None:
+    """Test listing public trades."""
+    public_trades = await set_up["client"].list_public_trades(
+        delivery_period=set_up["delivery_period"],
+        max_nr_trades=10,
+    )
+    assert len(public_trades) >= 0
+
+
+async def test_stream_gridpool_orders(set_up: dict[str, Any]) -> None:
+    """Test streaming gridpool orders."""
+    stream = await set_up["client"].stream_gridpool_orders(GRIDPOOL_ID)
+    await create_test_order(set_up)
+
+    try:
+        # Stream trades with a 15-second timeout to avoid indefinite hanging
+        streamed_order = await asyncio.wait_for(anext(stream), timeout=15)
+        assert streamed_order is not None, "Failed to receive streamed order."
```

Comment:
Are you expecting to match the exact order that you just created? I so, can we assert that?
Also, how likely is that this will hang? Any idea?

Yet another also, shouldn't we start streaming _before_ we create the order?

Reply:
Yes, I added an equality check now (I guess since no other gridpool is using this it works, but not gonna lie it feels a bit hacky to do it like this). The reason why the other is created before is because otherwise the code gets blocked at the line:  `streamed_order = await asyncio.wait_for(anext(stream), timeout=15) `. It works like this though but I guess it's because the order creation is fast enough to have the order received in the stream. So yea it works for the test but feels a bit hacky